### PR TITLE
Test cases for Promo code

### PR DESCRIPTION
**Pull request by islamov089**

### CODE REVIEW

File: src/helper/wrapper/playwrightWrappers.ts [hunk 1]
```diff
@@ -10,6 +12,8 @@ export default class PlaywrightWrapper {
   }
 
   async waitAndClick(locator: Locator) {
+    await this.page.waitForLoadState('networkidle');
```
Comment (jkalandarov) [inline src/helper/wrapper/playwrightWrappers.ts]: Official Playwright doc discourages using `'networkidle'` as a waiter: https://playwright.dev/docs/api/class-page#page-wait-for-load-state

File: src/pages/campaignDetails.ts
Comment (jkalandarov) [inline src/pages/campaignDetails.ts:1]: add `Page` to page object file name => CampaignDetailsPage.ts
Class names should be snake cases in js/ts.

File: src/pages/campaignDetails.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  constructor(private page: Page) {
+    this.base = new PlaywrightWrapper(page);
+  }
+  optionsButton = this.page.locator("(//button[contains(@class,'el-button button-ui--icon')]//i)[2]")
+  editButton = this.page.locator("//span[text()='Edit']")
+  checkoutButton = this.page.locator("(//button[contains(@class,'el-button el-button--primary')]//span)[2]")
+  addPromoCodeButton = this.page.locator("//button[contains(@class,'el-button button-ui--icon')]//i[1]")
+  promoCodeInput = this.page.locator("//div[contains(@class,'el-input el-input-group')]//input[1]")
+  applyButton = this.page.locator("//div[@class='el-input-group__append']//a[1]")
+  discountInfo = this.page.locator("(//span[@class='ui-money'])[3]")
```
Comment (jkalandarov) [inline src/pages/campaignDetails.ts]: not sure what element this locator is referring to. Can you work on naming the locator?

File: src/pages/promoCodePage.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  channelListDropdown = this.page.locator("(//span[@class='el-input__suffix-inner']//i)[3]");
+  discountTypeDropdown =this.page.locator("(//span[text()='Discount type *']/following::input)[2]")
+  promoCodeInput = this.page.locator("(//span[text()='Code *']/following::input)[1]")
+  startDateInput = this.page.locator("(//span[text()='Start date *']/following::input)[2]")
+  endDateInput = this.page.locator("(//span[text()='End date *']/following::input)[2]")
+  discountValueInput  = this.page.locator("(//span[text()='Discount value *']/following::input)[1]")
+  minBudgetInput =this.page.locator("(//span[text()='Min budget *']/following::input)[1]")
+  maxBudgetInput = this.page.locator("(//span[text()='Max budget *']/following::input)[1]")
+  currencyListDropdown = this.page.locator("(//span[text()='Currency *']/following::input)[2]")
+  theCodeCanBeUsed = this.page.locator("(//span[text()='The code can be used *']/following::input)[1]")
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: No need to use articles in variable/property names. what kind of element is `codeCanBeUsed`?

File: src/pages/promoCodePage.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  hoverLocator = this.page.locator("(//td[contains(@class,'el-table_1_column_7 ')]//div)[1]")
+  publishButton = this.page.locator("//span[text()=' Ok ']")
+
+  public generateUniquePromoCodeName() {
+    const currentDate = new Date();
+    const month = (currentDate.getMonth() + 1).toString().padStart(2, '0');
+    const day = currentDate.getDate().toString().padStart(2, '0');
+    const year = currentDate.getFullYear().toString().slice(-2);
+    const hour = currentDate.getHours().toString().padStart(2, '0');
+    const minute = currentDate.getMinutes().toString().padStart(2, '0');
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: You can create a separate util class that works with the Date object. You can create various methods inside that class. After all, you can encapsulate that class in your page object.

File: src/pages/promoCodePage.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  public setPromoCodeDate() {
+    let months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
+    let today = new Date();
+    let month = months[today.getMonth()];
+    let day = today.getDate();
+    let year = today.getFullYear();
+    return month + ' ' + day + ', ' + year;
+  }
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: This method merely creates a string in some date format. Therefore it should belong to a different class.

File: src/pages/promoCodePage.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  public async hoverOverElementAndClickButton(): Promise<void> {
+    await this.hoverLocator.waitFor();
+    await this.hoverLocator.hover();
+    await this.activeButton.waitFor();
+    await this.activeButton.click();
+  }
+
+  public selectOption(option: string): Locator {
+    return  this.page.locator(`(//span[text()='${option}'])`)
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: this method actually doesn't select anything but returns a locator.

File: src/pages/promoCodePage.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  }
+  public promoCodeCurrency(option: string): Locator {
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: method name should have a verb, and action in its implementation block.

File: src/pages/promoCodePage.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  }
+}
+export const { getPromoCode, setPromoCode } = (function promoCodeCache() {
+  let promoCode = null;
+
+  return {
+    setPromoCode: (newPromoCode: string | null) => (promoCode = newPromoCode),
+    getPromoCode: () => promoCode
+  };
+})()
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: I am confused as to why you are exporting `getPromoCode` and `setPromoCode` variables from here and declaring them by using IIFE.

File: src/test/features/createPromoCode.feature [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+        Scenario: Create and check Promo code
+             When User logins as a "admin"
+             Then User is in "Dashboard" page
+             When User navigates to "Settings Tab" from "Dashboard Page"
+              And User clicks "Promo Codes Tab" on "Settings Page"
+              And User clicks "New Promo Code Button" on "Promo Code Page"
+              And User selects "GAM Network" from "Channel List Dropdown" on "Promo Code Page"
+              And User named Promo code to "Promo Code input" on "Promo Code Page"
+              And User select dates to "Start Date Input" and "End Date Input" on "Promo Code Page"
```
Comment (jkalandarov) [inline src/test/features/createPromoCode.feature]: user selects which dates?

File: src/test/features/createPromoCode.feature [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+              And User fills "Discount Value Input" as "10" on "Promo Code Page"
+              And User fills "Min Budget Input" as "1" on "Promo Code Page"
+              And User fills "Max Budget Input" as "999" on "Promo Code Page"
+              And User select "USD" from "Currency List Dropdown" on "Promo Code Page"
+              And User fills "The code can be used" as "1" on "Promo Code Page"
+              And User clicks "Save Button" on "Promo Code Page"
+              And User clicks "Publish button" for activate Promo Code on "Promo Code Page"
+              And User navigates to "Dashboard Tab" from "Dashboard Page"
+              And User clicks "All Button" on "Dashboard Page"
+              And User fills "Search Input" as "Test promo code" on "Dashboard Page"
```
Comment (jkalandarov) [inline src/test/features/createPromoCode.feature]: Do you create `Test promo code` campaign manually before this scenario?

**Review by jkalandarov — CHANGES_REQUESTED** [review]
Pay attention to code style rules, especially spaces.

File: src/pages/promoCodePage.ts (hunk 8 above)
Comment (islamov089) [reply src/pages/promoCodePage.ts]: I wanted to store  Promo code state and use it in another file.

File: src/pages/promoCodePage.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  maxBudgetInput = this.page.locator("(//span[text()='Max budget *']/following::input)[1]");
+  currencyListDropdown = this.page.locator("(//span[text()='Currency *']/following::input)[2]");
+  codeCanBeUsedInput = this.page.locator("(//span[text()='The code can be used *']/following::input)[1]");
+  saveButton = this.page.locator("//span[text()=' Save ']");
+  activeButton = this.page.locator("(//i[@class='icon-play'])[1]");
+  hoverLocator = this.page.locator("(//td[contains(@class,'el-table_1_column_7 ')]//div)[1]");
+  publishButton = this.page.locator("//span[text()=' Ok ']");
+
+
+  public async hoverOverElementAndClickButton(): Promise<void> {
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: The method name doesn't say on which button this method hovers over and clicks. I assume it should do those actions on any buttons, but the method doesn't accept parameters. So, It is confusing. You should, etheir refactor this method, or rename it.

File: src/pages/promoCodePage.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    await this.hoverLocator.waitFor();
+    await this.hoverLocator.hover();
+    await this.activeButton.waitFor();
+    await this.activeButton.click();
+  }
+  public generateUniquePromoCodeName() {
+    return `AU${new Date().valueOf().toString().substring(5)}`
+  }
+  selectedPromoCodeDates(){
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: space between

File: src/test/features/generateReport.feature
Comment (jkalandarov) [inline src/test/features/generateReport.feature:1]: Why did you change this file? Is this change included in the scope of your task?

File: src/test/steps/generateReportSteps.ts
Comment (jkalandarov) [inline src/test/steps/generateReportSteps.ts:1]: Pay attention you are making unnecessary changes which is extra work for code reviewer.

**Review by jkalandarov — CHANGES_REQUESTED** [review]

File: src/pages/promoCodePage.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  minBudgetInput = this.page.locator("(//span[text()='Min budget *']/following::input)[1]");
+  maxBudgetInput = this.page.locator("(//span[text()='Max budget *']/following::input)[1]");
+  currencyListDropdown = this.page.locator("(//span[text()='Currency *']/following::input)[2]");
+  codeCanBeUsedInput = this.page.locator("(//span[text()='The code can be used *']/following::input)[1]");
+  saveButton = this.page.locator("//span[text()=' Save ']");
+  activeButton = this.page.locator("(//i[@class='icon-play'])[1]");
+  hoverLocator = this.page.locator("(//td[contains(@class,'el-table_1_column_7 ')]//div)[1]");
+  publishButton = this.page.locator("//span[text()=' Ok ']");
+
+  public async hoverOverAndClickPayButton(): Promise<void> {
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: Does this method click on `Activate` button or `Pay` button?

File: src/pages/promoCodePage.ts [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    await this.hoverLocator.hover();
+    await this.activeButton.waitFor();
+    await this.activeButton.click();
+  }
+
+  public generateUniquePromoCodeName() {
+    return `AU${new Date().valueOf().toString().substring(5)}`;
+  }
+
+  selectedPromoCodeDates() {
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: The method doesn't select any or perform any action. It just returns a string. Try to find a suitable name for the method.

File: src/pages/signUpPage.ts [hunk 15]
```diff
@@ -14,15 +14,15 @@ export default class SignUpPage {
   countryListDropdown = this.page.locator("(//i[contains(@class,'el-select__caret el-input__icon')])[1]");
   currencyListDropdown = this.page.locator("(//i[contains(@class,'el-select__caret el-input__icon')])[2]");
   industryListDropdown = this.page.locator("(//i[contains(@class,'el-select__caret el-input__icon')])[3]");
-  phoneNumberInput = this.page.locator("(//span[text()='Phone']/following::input)[1]");
+  // phoneNumberInput = this.page.locator("(//span[text()='Phone']/following::input)[1]");
```
Comment (jkalandarov) [inline src/pages/signUpPage.ts]: Why did you comment the locator?

File: src/test/features/generateReport.feature
Comment (jkalandarov) [inline src/test/features/generateReport.feature:1]: Is there any specific reason for changing this file?

**Review by jkalandarov — CHANGES_REQUESTED** [review]
Pay attention to previous comments. You've not changed anything but only linted some files.

**Review by jkalandarov — COMMENTED** [review]
Oldingi yozgan kommentariyalarimga etibor bering!!!

File: src/pages/promoCodePage.ts [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  public generateUniquePromoCodeName() {
+    return `AU${new Date().valueOf().toString().substring(5)}`;
+  }
+
+  setPromoCodeDates() {
+    const formattedToday = dataUtil.getTodayFormatted();
+    return formattedToday;
+  }
```
Comment (jkalandarov) [inline src/pages/promoCodePage.ts]: Please look at my previous comment regarding this method.

File: src/test/steps/createPromoCodeSteps.ts [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+When('User named Promo code to {string} on {string}', async function (input, page) {
+  const promoCodeName = basePage.promoCodePage.generateUniquePromoCodeName();
+  const [codeInput, pageName] = basePage.wrapper.toCamelCase(input, page);
+  await setPromoCode(promoCodeName);
+  await basePage.wrapper.type(basePage[pageName][codeInput], promoCodeName);
+});
+
+When('User selects 1 day for Promo Code duration to {string} and {string} on {string}', async function (firstInput, secondInput, page) {
+  const datesInput = basePage.promoCodePage.setPromoCodeDates();
```
Comment (jkalandarov) [inline src/test/steps/createPromoCodeSteps.ts]: `setPromoCodeDates` returns a string. `datesInput` is a confusing name here.

File: src/test/steps/createPromoCodeSteps.ts [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+  const [startDayInput, endDateInput, pageName] = basePage.wrapper.toCamelCase(firstInput, secondInput, page);
+  await basePage.wrapper.type(basePage[pageName][startDayInput], datesInput);
+  await basePage.wrapper.type(basePage[pageName][endDateInput], datesInput);
```
Comment (jkalandarov) [inline src/test/steps/createPromoCodeSteps.ts]: `wrapper.type` accepts two parameters. First one is the input name, second is the text to be typed into the input. You are passing inputs to both parameters.

**Review by jkalandarov — CHANGES_REQUESTED** [review]
Ilhom! I've repeated at least 100 times to be attentive. You've been ignoring my comments in every code review. What's the point of conducting a code review in that case?

**Comment by islamov089** [comment]
@jkalandarov aka campaignWizard.tsni delete qilaman deb, campaignWizardPage.ts ni delete qivorgan ekanman boya to'girlab qo'ydim.

**Review by jkalandarov — APPROVED** [review]
Good job!

**Pull request merged**